src/lib.rs

```diff
@@ -37,13 +37,9 @@ impl Nes {
     unsafe fn map_devices(self: &mut Pin<Box<Self>>) {
         let pinned_nes_ref: Pin<&mut Self> = Pin::as_mut(self);
         let nes_ref: &mut Self = Pin::get_unchecked_mut(pinned_nes_ref);
-
-        nes_ref.joypad_1.map_memory(&mut nes_ref.memory[0x4016]);
-        nes_ref.joypad_2.map_memory(&mut nes_ref.memory[0x4017]);
-
-        nes_ref.bus.joypad_1 = &mut nes_ref.joypad_1;
-        nes_ref.bus.joypad_2 = &mut nes_ref.joypad_2;
-        
-        nes_ref.cpu.bus = &mut nes_ref.bus;
+        nes_ref.bus.map(
+            &mut nes_ref.memory,
+            &[&mut nes_ref.joypad_1, &mut nes_ref.joypad_2],
+        )
```

Comment:
addresses are part of the mapping, so it's weird that we pass them at construction, but the actual mapping is done later on

Reply:
Well here the idea is that the devices proves it's own definition of the memory range it handles. When we ask the bus to make the mapping with provided devices, it's give each device the memory slice it's concerned with.